### PR TITLE
Improve reducer

### PR DESCRIPTION
**Pull request by zurgl**
- A new **ProtocolNav** component generalize and factorize the nav at protocol level
- **SetInnerState** reducer done as describe on the card
- **SetSharedState** reducer done as describe on the card
- Keep **SetStepCompleted** for convenience usage
- All the read-only method have been migrated to **utils/context.ts**
- Error message management have been externalized to **utils/error.ts**
- All the function name have been simplified to increase readabilities
- All future removed function have been marked as deprecated
- Some work need to be done to avoid confusion between NETWORKS and PROVIDER (currently named NETWORK)

Screen gif as a visual proof that's working fine
solana-new

### CODE REVIEW

**Comment by zurgl** [comment]
we don't see it on the gif buit there is a divider between the ProtocolNav and the step component

File: context/index.tsx [hunk 1]
```diff
@@ -21,20 +28,21 @@ type StepsReducerHelperT = {
 const stepsReducerHelper = (
   {index, data}: {index: number; data: ProtocolStepsT},
   step: StepType,
-): StepsReducerHelperT => {
+): {
+  index: number;
+  data: ProtocolStepsT;
+} => {
```
Comment (guillaumegaluz) [inline context/index.tsx:34]: reuse `StepsReducerHelperT`

File: context/index.tsx [hunk 2]
```diff
@@ -129,10 +168,122 @@ export type Action =
...
+        },
+      };
+    }
+
+    case 'SetInnerState': {
+      const chainId = getChainId(state);
+      const stepId = getStepId(state);
+      const isCompleted = action.isCompleted ? true : isCompletedStep(state);
+      const innerState = state.protocols[chainId].innerState;
+      console.log(action.values);
```
Comment (guillaumegaluz) [inline context/index.tsx:205]: remove

File: context/index.tsx [hunk 3]
```diff
@@ -129,10 +168,122 @@ export type Action =
...
+          },
+        },
+      };
+    }
+
+    case 'SetSharedState': {
+      const chainId = getChainId(state);
+      const stepId = getStepId(state);
+      const isCompleted = action.isCompleted ? true : isCompletedStep(state);
+      let protocolState = {...state.protocols[chainId]};
```
Comment (guillaumegaluz) [inline context/index.tsx:239]: `let sharedState`

File: types/index.ts [hunk 4]
```diff
@@ -158,6 +163,13 @@ export type NETWORKS =
   | TEZOS_NETWORKS
   | ARWEAVE_NETWORKS;
 
+// -----------------------------
+export enum NETWORK {
+  DATAHUB,
+  TESTNET,
+  LOCALNET,
```
Comment (guillaumegaluz) [inline types/index.ts:170]: not in this PR, but in another one, can we change to:

```
LOCAL
TESTNET
TESTNET_VIA_DH
```

File: utils/index.ts [hunk 5]
```diff
@@ -1,3 +1,5 @@
+export const VERSION = 'DEPRECATED';
```
[the remark this answers is not in the document]
Comment (zurgl) [reply utils/index.ts]: a file I need to remove. Done

File: components/protocols/ceramic/components/auth/index.tsx [hunk 6]
```diff
@@ -190,6 +190,7 @@ const Auth = (props: AuthProps): JSX.Element => {
               onClick={handleDisconnect}
               block
               danger
+              key={1}
```
[the remark this answers is not in the document]
Comment (zurgl) [reply components/protocols/ceramic/components/auth/index.tsx]: ok

File: components/protocols/solana/components/Balance.tsx [hunk 7]
```diff
@@ -1,39 +1,35 @@
...
+import {ErrorBox, ErrorT, prettyError} from 'utils/error';
 import {useEffect, useState} from 'react';
-import {
-  getCurrentChainId,
-  useGlobalState,
-  getCurrentStepIdForCurrentChain,
-  getNetworkForCurrentChain,
-  getChainInnerState,
-} from 'context';
+import {useGlobalState, getNetworkForCurrentChain} from 'context';
```
[the remark this answers is not in the document]
Comment (zurgl) [reply components/protocols/solana/components/Balance.tsx]: removed

**Comment by zurgl** [comment]
@guillaumegaluz Your feedback have been integrated to the branch

**Review by silentlight — APPROVED** [review]
Looks good to me although it is pretty complex :)

**Pull request merged**